### PR TITLE
New diags also handle particle

### PR DESCRIPTION
**Pull request by MaxThevenet**
Similar to what was done for fields in new diagnostics, but for particles: the diagnostics handling is moved from `WarpXParticleContainer` to a separate class `ParticleDiag` in `Source/Diagnostics/`. Besides doing some refactoring, this adds the capability to have different particle output for different instances of diagnostics (more clearly, you can write all particles of all species every 1000 iterations, and only `x y z ux uz By` of 1% of `electrons` every 10 iterations, for instance).

Right now, this is working for plotfiles only, still work to be done to have openPMD output working. 

Again, the old diagnostics have not been deleted, and the doc is not updated, this will be done in a future PR.

@atmyers @Yin-YinjianZhao I had to modify a bit the parser filter for this. Could you check if this still makes sense?

### CODE REVIEW

**Comment by MaxThevenet** [comment]
Update: what's currently in this PR is ready for review, but I'll also try to update OpenPMD output, and if it's not too much change I'll include it in this PR.

File: Source/Diagnostics/ParticleDiag/ParticleDiag.cpp [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+{
+    ParmParse pp(diag_name + "." + name);
+    if (!pp.queryarr("variables", variables)){
+        variables = {"Ex", "Ey", "Ez", "Bx", "By", "Bz", "ux", "uy", "uz", "w"};
+    }
+
+    //variable to set plot_flags size
+    int plot_flag_size = PIdx::nattribs;
+    if(WarpX::do_back_transformed_diagnostics && m_pc->doBackTransformedDiagnostics())
+        plot_flag_size += 6;
```
Comment (ax3l) [inline Source/Diagnostics/ParticleDiag/ParticleDiag.cpp:20]: maybe comment which additional attributes are counted here

Comment (ax3l) [inline Source/Diagnostics/ParticleDiag/ParticleDiag.cpp:20]: maybe comment which additional attributes are counted here

File: Source/Diagnostics/ParticleDiag/ParticleDiag.cpp [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+#ifdef WARPX_DIM_RZ
+    // Always write out theta, whether or not it's requested,
+    // to be consistent with always writing out r and z.
+    plot_flags[ParticleStringNames::to_index.at("theta")] = 1;
+#endif
+
+#ifdef WARPX_QED
+    if(m_pc->DoQED()){
+        //Optical depths is always plotted if QED is on
+        plot_flags[plot_flag_size-1] = 1;
```
Comment (ax3l) [inline Source/Diagnostics/ParticleDiag/ParticleDiag.cpp:54]: this looks dangerous, if someone adds another attribute.
would a `map` work better for `plot_flags`?

File: Source/Diagnostics/WarpXOpenPMD.cpp [hunk 3]
```diff
@@ -255,10 +254,71 @@ WarpXOpenPMDPlot::WriteOpenPMDParticles (MultiParticleContainer& mpc)
...
+    }
+
+    // Convert momentum to SI
+    pc->ConvertUnits(ConvertDirection::WarpX_to_SI);
+    // real_names contains a list of all real particle attributes.
+    // pc->plot_flags is 1 or 0, whether quantity is dumped or not.
+
+    {
+      //
+      SavePlotFile(pc,
```
Comment (ax3l) [inline Source/Diagnostics/WarpXOpenPMD.cpp:306]: feel free to rename this function, it's just confusing but we kept it b/c it looked similar to the blueprint we started with.

File: Source/Diagnostics/ParticleDiag/ParticleDiag.cpp (hunk 1 above)
Comment (MaxThevenet) [reply Source/Diagnostics/ParticleDiag/ParticleDiag.cpp:20]: Good point, I'l add a comment in the next commit.

File: Source/Diagnostics/ParticleDiag/ParticleDiag.cpp (hunk 2 above)
Comment (MaxThevenet) [reply Source/Diagnostics/ParticleDiag/ParticleDiag.cpp:54]: Thanks for pointing it out, I agree it is misleading, dangerous and also probably wrong (if both QSR and BW are on, I expect this to cause a bug). A solution would be to do the same as for other variables above in the same file, e.g.,
```c++
ParticleStringNames::to_index.at("opt_depth_BW")
ParticleStringNames::to_index.at("opt_depth_breit_QSR")
```
but I do not think these exist. So I think the problem is general and should be fixed in the particle containers first. I can open an issue for that.

File: Source/Diagnostics/WarpXOpenPMD.cpp (hunk 3 above)
Comment (MaxThevenet) [reply Source/Diagnostics/WarpXOpenPMD.cpp:306]: Sure, I renamed this to `DumpToFile`. Thanks!

**Comment by MaxThevenet** [comment]
Thanks @Yin-YinjianZhao!

File: Source/Diagnostics/ParticleDiag/ParticleDiag.cpp (hunk 2 above)
Comment (ax3l) [reply Source/Diagnostics/ParticleDiag/ParticleDiag.cpp:55]: sounds good, can you just add a TODO or warning note here, so one can grep for this?

File: Source/Diagnostics/ParticleDiag/ParticleDiag.cpp [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+#ifdef WARPX_DIM_RZ
+    // Always write out theta, whether or not it's requested,
+    // to be consistent with always writing out r and z.
+    plot_flags[ParticleStringNames::to_index.at("theta")] = 1;
+#endif
+
+#ifdef WARPX_QED
+    if(m_pc->DoQED()){
+        //Optical depths is always plotted if QED is on
```
Comment (MaxThevenet) [inline Source/Diagnostics/ParticleDiag/ParticleDiag.cpp:54]:
```suggestion
        // TODO: cleaner handling of particle attributes. This should probably be done in
        // the particle container classes, either by adding "opt_depth_BW" to ParticleStringNames
        // or by using a std::map for particle attributes.
        // Optical depths is always plotted if QED is on.
```

File: Source/Diagnostics/WarpXOpenPMD.cpp [hunk 5]
```diff
@@ -255,10 +254,71 @@ WarpXOpenPMDPlot::WriteOpenPMDParticles (MultiParticleContainer& mpc)
   }
 }
 
+void
+WarpXOpenPMDPlot::WriteOpenPMDParticles (const amrex::Vector<ParticleDiag>& particle_diags)
+{
+  WARPX_PROFILE("WarpXOpenPMDPlot::WriteOpenPMDParticles()");
+  //std::vector<std::string> species_names = mpc.GetSpeciesNames();
```
Comment (ax3l) [inline Source/Diagnostics/WarpXOpenPMD.cpp:261]:
```suggestion
```

File: Source/Diagnostics/FlushFormats/FlushFormatPlotfile.cpp [hunk 6]
```diff
@@ -42,11 +43,12 @@ FlushFormatPlotfile::WriteToFile (
     WriteAllRawFields(plot_raw_fields, nlev, filename, plot_raw_fields_guards,
                       plot_raw_rho, plot_raw_F);
 
-    mpc.WritePlotFile(filename);
+    //mpc.WritePlotFile(filename);
```
Comment (ax3l) [inline Source/Diagnostics/FlushFormats/FlushFormatPlotfile.cpp:46]:
```suggestion
```

File: Source/Diagnostics/ParticleDiag/ParticleDiag.cpp [hunk 7]
```diff
@@ -0,0 +1,75 @@
+#include "WarpX.H"
+#include "ParticleDiag.H"
+// #include "Utils/WarpXUtil.H"
```
Comment (RevathiJambunathan) [inline Source/Diagnostics/ParticleDiag/ParticleDiag.cpp:3]: this can be deleted.

File: Source/Diagnostics/WarpXOpenPMD.H [hunk 8]
```diff
@@ -156,7 +159,7 @@ private:
    * @param[in] write_int_comp The int attribute ids, from WarpX
    * @param[in] int_comp_names The int attribute names, from WarpX
    */
-  void SavePlotFile(const std::unique_ptr<WarpXParticleContainer>& pc,
+  void DumpToFile(WarpXParticleContainer* pc,
```
Comment (RevathiJambunathan) [inline Source/Diagnostics/WarpXOpenPMD.H:162]: is the commented out code on Line 173 needed? it was not part of this PR, so may be can done somewhere else. But just wanted to note it here.

File: Source/Diagnostics/WarpXOpenPMD.cpp [hunk 9]
```diff
@@ -255,10 +254,70 @@ WarpXOpenPMDPlot::WriteOpenPMDParticles (MultiParticleContainer& mpc)
...
+       // dumped as particle record in a plotfile. So far, ionization_level is the only
+       // integer attribs, and it is automatically dumped as particle record
+       // when ionization is on.
+       int_flags.resize(1, 1);
+    }
+
+    // Convert momentum to SI
+    pc->ConvertUnits(ConvertDirection::WarpX_to_SI);
+    // real_names contains a list of all real particle attributes.
+    // pc->plot_flags is 1 or 0, whether quantity is dumped or not.
```
Comment (RevathiJambunathan) [inline Source/Diagnostics/WarpXOpenPMD.cpp:301]: in the comment : pc->plot_flags will be particle_diags[i].plot_flags, since that is what is called below.

File: Source/Diagnostics/WarpXOpenPMD.cpp [hunk 10]
```diff
@@ -255,10 +254,70 @@ WarpXOpenPMDPlot::WriteOpenPMDParticles (MultiParticleContainer& mpc)
...
+
+    {
+      //
```
Comment (RevathiJambunathan) [inline Source/Diagnostics/WarpXOpenPMD.cpp:304]: // can be deleted.

File: Source/Particles/Filter/FilterFunctors.H [hunk 11]
```diff
@@ -111,7 +108,7 @@ struct ParserFilter
     }
 public:
     amrex::Real m_t;
-    ParserWrapper<7> * m_function_partparser;
+    ParserWrapper<7> const * const m_function_partparser;
```
Comment (RevathiJambunathan) [inline Source/Particles/Filter/FilterFunctors.H:111]: please can you add a comment explaining the number 7.

File: Source/Particles/MultiParticleContainer.H [hunk 12]
```diff
@@ -59,9 +59,11 @@ public:
...
     ~MultiParticleContainer() {}
 
-    WarpXParticleContainer& GetParticleContainer (int ispecies) {
-        return *allcontainers[ispecies];
-    }
+    WarpXParticleContainer&
+    GetParticleContainer (int ispecies) const {return *allcontainers[ispecies];}
+
+    WarpXParticleContainer*
+    GetParticleContainerPtr (int ispecies) const {return allcontainers[ispecies].get();}
```
Comment (RevathiJambunathan) [inline Source/Particles/MultiParticleContainer.H:66]: Do we use both functions?

Comment (MaxThevenet) [reply Source/Particles/MultiParticleContainer.H:66]: Yes, one in the old diags, one in the new. One will be deleted in another PR, when we get rid of the old ones.

**Comment by MaxThevenet** [comment]
Thanks for your suggestions! They should all be included in the last commit.

**Comment by MaxThevenet** [comment]
\o/
